compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -431,45 +741,204 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
+                (AggregateTy::Tuple, FIRST_VARIANT)
+            }
+            AggregateKind::Closure(did, substs) | AggregateKind::Generator(did, substs, _) => {
+                (AggregateTy::Def(did, substs), FIRST_VARIANT)
+            }
+            AggregateKind::Adt(did, variant_index, substs, _, None) => {
+                (AggregateTy::Def(did, substs), variant_index)
+            }
+            // Do not track unions.
+            AggregateKind::Adt(_, _, _, _, Some(_)) => return None,
+        };
+
+        let fields: Option<Vec<_>> = fields
+            .iter_mut()
+            .map(|op| self.simplify_operand(op, location).or_else(|| self.new_opaque()))
+            .collect();
+        let fields = fields?;
+
+        if let AggregateTy::Array = ty && fields.len() > 4 {
+            let first = fields[0];
+            if fields.iter().all(|&v| v == first) {
+                let len = ty::Const::from_target_usize(self.tcx, fields.len().try_into().unwrap());
+                if let Some(const_) = self.try_as_constant(first) {
+                    *rvalue = Rvalue::Repeat(Operand::Constant(Box::new(const_)), len);
+                } else if let Some(local) = self.try_as_local(first, location) {
+                    *rvalue = Rvalue::Repeat(Operand::Copy(local.into()), len);
+                    self.reused_locals.insert(local);
+                }
+                return Some(Value::Repeat(first, len));
+            }
+        }
+
+        let value = Value::Aggregate(ty, variant_index, fields);
+        Some(value)
+    }
+}
+
+fn op_to_prop_const<'tcx>(
+    ecx: &mut InterpCx<'_, 'tcx, DummyMachine>,
+    op: &OpTy<'tcx>,
+) -> Option<ConstValue<'tcx>> {
+    // Do not attempt to propagate unsized locals.
+    if op.layout.is_unsized() {
+        return None;
+    }
+
+    // This constant is a ZST, just return an empty value.
+    if op.layout.is_zst() {
+        return Some(ConstValue::ZeroSized);
+    }
+
+    // Do not synthetize too large constants. Codegen will just memcpy them, which we'd like to avoid.
+    if !matches!(op.layout.abi, Abi::Scalar(..) | Abi::ScalarPair(..)) {
+        return None;
+    }
+
+    // If this constant has scalar ABI, return it as a `ConstValue::Scalar`.
+    if let Abi::Scalar(abi::Scalar::Initialized { .. }) = op.layout.abi
+        && let Ok(scalar) = ecx.read_scalar(op)
+        && scalar.try_to_int().is_ok()
+    {
+        return Some(ConstValue::Scalar(scalar));
+    }
+
+    // If this constant is a projection of another, we can return it directly.
+    if let Either::Left(mplace) = op.as_mplace_or_imm()
+        && let MemPlaceMeta::None = mplace.meta()
+    {
+        let (size, align) = ecx.size_and_align_of_mplace(&mplace).ok()??;
+
+        let alloc_ref = ecx.get_ptr_alloc(mplace.ptr(), size, align).ok()??;
+        // Do not try interning a value that contains provenance.
+        if alloc_ref.has_provenance() {
+            return None;
+        }
```

Comment:
Seems nicer to let `intern_const_alloc_for_constprop` do this check, which naturally has access to the allocation anyway.

Reply:
We definitely need to handle some pointers. For instance when accessing a static to read from it.
What we cannot do is write pointers in the resulting MIR, because codegen does not respect identity of alloc-id in a across crates.

Reply:
But here this hasn't about "handling" pointers, it's about the constants that are actually tracked as "numbered" expressions and written back into the MIR, right? Can those ever contain pointers?

Reply:
Did you change your mind from https://github.com/rust-lang/rust/pull/116270#discussion_r1349748917? I can revert https://github.com/rust-lang/rust/pull/116270/commits/ac4572f7bfbd4ad6bf35c85814c96c101e4319f9

Reply:
The practical issue I have is this one:
```
static A: fn() = || {};

// In another crate.
fn foo() -> fn() {
    A
}

// In another crate.
fn bar() {
  assert_eq!(A, foo());
}
```

Consider:
- `A` is at address `alloc1`;
- `*alloc1` is the pointer `alloc2`;
- if the opt writes `alloc2` in the MIR for `foo`, the assertion in `bar` may fail because codegen does not guarantee equality of fn pointers, which is a miscompilation.

On the other hand, for
```
static B: &'static &'static u8 = &&63;

fn blah() -> u8 {
    **A
}
```

We definitely want to optimize `blah` to return 63.

Reply:
> But here this hasn't about "handling" pointers, it's about the constants that are actually tracked as "numbered" expressions and written back into the MIR, right? Can those ever contain pointers?

In short, yes. They can.

Reply:
> They can contain pointers when being numbered, but will not be written back if they do.

Okay, that was my wrong assumption then. So for `**A` to be turned into 63, we need to number `A` and `*A`?

However, wouldn't we want to also optimize `*A` to be a pointer directly into the allocation that contains `&63`? That should also be sound, right? The issue isn't pointers in general, it's pointers created by certain operations: consts, and fn-item-to-fn-ptr coercions. So the more principled approach seems to be to never number fn-item-to-fn-ptr coercions, and when encountering a const to check whether it contains any provenance before numbering it. Once we do that any provenance we still number is fine?

Though I'm not entirely sure about statics with weak linkage. So maybe the point is you want to make a wide berth around this entire problem area even if that's unnecessarily conservative?

Either way this needs comments in the code that explain the why and what.

Reply:
Fundamentally, the problem is introducing more non-determinism. So when we write back we must not write back something that could have non-deterministic choice where there was no non-deterministic choice before.

What are the actually non-deterministic operations? An unevaluated constant can have non-determinism, so we can never replace `x` by `C`. However, we *can* replace `x` by the result obtained from evaluating `C`! An evaluated constant doesn't have extra non-determinism in general. So is this one even a problem for GVN, does GVN track unevaluated constants symbolically? If no, there should probably be a comment somewhere explaining why that would be a problem.

And then there's function pointers. Those are nasty indeed since even the same evaluated result can lead to different values at runtime, I think. The way to protect against that, I think, is to never number a value that contains provenance *that points to a function allocation* (or a vtable allocation... I'm not sure if we currently mark them as "address insignificant" but it seems like a reasonable thing to do).

If we do that, then we ensure that bad values never enter the GVN system, and then everything that we *do* number can be put back into the program, right?

Reply:
You are right. I found an example of erroneous merging of constants.
```rust
fn opaque<T>(x: T) -> T { x }

#[custom_mir(dialect="analysis")]
fn foo() -> (bool, bool) {
    mir!(
        let au: u128;
        let bu: u128;
        let cu: u128;
        let du: u128;
        let c: &str;
        let d: &str;
        {
            let a = ("a",);
            Call(au = transmute::<_, u128>(a.0), bb1)
        }
        bb1 = {
            Call(c = opaque(a.0), bb2)
        }
        bb2 = {
            Call(cu = transmute::<_, u128>(c), bb3)
        }
        bb3 = {
            let b = "a";
            Call(bu = transmute::<_, u128>(b), bb4)
        }
        bb4 = {
            Call(d = opaque(b), bb5)
        }
        bb5 = {
            Call(du = transmute::<_, u128>(d), bb6)
        }
        bb6 = {
            let direct = au == bu;
            let indirect = cu == du;
            RET = (direct, indirect);
            Return()
        }
    )
}
```

Normally, both `direct` and `indirect` hold the same value. The current implementation merges `au` and `bu`, so `direct` is always true, but fails to merge `a.0` and `b` so `cu` and `du` may be different. Miri fails when enabling GVN on this case.

Reply:
So specifically what is happening here is numbering of unevaluated consts, with subsequent duplication of the non-determinism that arises from their evaluation?

Reply:
Yes. We give the same VnIndex to `a.0` and `b`, although miri/codegen is allowed to give them different AllocIds. The latest push adds a disambiguation mechanism for non-deterministic constants.